Fix aggressive error out behavior

byucesoy opened the pull request:
Fixes #1422 
Fixes #1424 

With this PR we made two main changes each in their respective commits.

We first fixed the crash occurring upgrading Citus 5.2 to Citus 6.2. To do that we remove ActivateNode call in master_initialize_node_metadata. In fact, we did not need to activate the nodes added via master_initialize_node_metadata anyway. Because this function is called **once** at either;

* upgrading from 5.2 to 6.0 and there are no reference tables or isactive column in Citus 5.2

or

* during extension creation and there are no reference tables created at that moment

The second commit is to only error out for distributed queries in case of version mismatch. Previously, we errored out on almost all kinds of queries because when there is a version mismatch it is not safe to use distributed cache which expects metadata tables in particular structure. With this change, we added a function to read pg_dist_partition metadata table directly without using distributed cache. In this function, we try to minimize our assumptions about pg_dist_partition. We only expect the name of the table, the position of the logicalrelid column and the name of the index on the logicalrelid column stays same across versions.

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -194,13 +195,68 @@ IsDistributedTable(Oid relationId)
+	 * specific details which are most likely do not change in long term.
+	 */
+	installedVersion = InstalledExtensionVersion();
+	if (!MajorVersionsCompatible(installedVersion, CITUS_EXTENSIONVERSION))
```
This doesn't look good to me, InstalledExtensionVersion() is far from cheap.  I think you're going to have to integrate this with the state kept in CitusHasBeenLoaded(), i.e. once determined that things are compatible, it needs to be cached.

byucesoy replied:
InstalledExtensionVersion already caches the result and we set it to NULL in case of [CREATE/ALTER EXTENSION](https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/metadata_cache.c#L1072)

anarazel replied:
Nevertheless, this is doing string parsing, still makes the function a lot slower in the common cached case.

byucesoy replied:
I cached the results of MajorVersionsCompatible :+1:

byucesoy commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -185,23 +184,29 @@ multi_ProcessUtility(Node *parsetree,
 	Oid savedUserId = InvalidOid;
 	int savedSecurityContext = 0;
 	List *ddlJobs = NIL;
-	bool skipCitusProcessing = SkipCitusProcessingForUtility(parsetree);
```
I removed this function, because in the version we error out almost all queries, we added this to whitelist some of the DDLs like CREATE/ALTER/DROP extension... Now since we do not error out non-distributed queries, we do not need this.

byucesoy commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -185,23 +184,29 @@ multi_ProcessUtility(Node *parsetree,
+	bool checkExtensionVersion = false;
 
-	if (skipCitusProcessing)
+	if (IsA(parsetree, TransactionStmt))
```
This check was originally here but we moved it to SkipCitusProcessingForUtility when we implemented it. I added it back.

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -194,13 +195,68 @@ IsDistributedTable(Oid relationId)
+										DistPartitionLogicalRelidIndexId(),
+										indexOK, NULL, scanKeyCount, scanKey);
+
+	partitionTuple = systable_getnext(scanDescriptor);
```
I'm a bit doubtful that it's a good idea to rely on DistPartitionRelationId(), DistPartitionLogicalRelidIndexId() etc here. But we can possibly defer that concern to a later point.

byucesoy commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -301,6 +301,9 @@ ConvertToDistributedTable(Oid relationId, char *distributionColumnName,
 	Var *distributionColumn = NULL;
 
+	/* we will not continue if there is a version mismatch */
+	ErrorIfInstalledVersionMismatch();
```
If there is a version mismatch and the table is distributed, we error out at IsDistributedTable and IsDistributedTable covers most of the cases. But here we also should error out of table is not distributed.

anarazel replied:
This is fairly late though, at that point we'll already have written to the catalog and such (which could cause issues w/ a mismatch).  That's not sufficient - I think this needs to be around the places where we currently check EnsureCoordinator();

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -435,28 +491,14 @@ DistributedTableCacheEntry(Oid distributedRelationId)
-	 * here.
-	 */
-	if (!CitusHasBeenLoaded())
+	if (!IsDistributedTable(distributedRelationId))
```
This *doubles* the lookup cost in the common case that this actually is a distributed table. Again, I don't think we should do this for the normal case.

anarazel commented:
Oh, and I'd argue this should add some tests to multi_extension.sql.

anarazel commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -301,6 +301,10 @@ ConvertToDistributedTable(Oid relationId, char *distributionColumnName,
 
+	/* we will not continue if there is a version mismatch */
+	ErrorIfAvailableVersionMismatch();
+	ErrorIfInstalledVersionMismatch();
```
I think your previous approach of combining these into one function made sense.

anarazel commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -111,6 +111,10 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 
+	/* we will not continue if there is a version mismatch */
+	ErrorIfAvailableVersionMismatch();
+	ErrorIfInstalledVersionMismatch();
```
Please combine into one wrapper function.

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -1119,30 +1205,101 @@ ErrorIfAvailableVersionMismatch(void)
+	/* if we cached the result before, return it */
+	if (availableVersionCompatible != NULL)
+	{
+		return *availableVersionCompatible;
```
Type mismatch.

byucesoy replied:
Umm, how? Anyway I will change this to regular bool, not bool pointer.

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -1119,30 +1205,101 @@ ErrorIfAvailableVersionMismatch(void)
+	oldMemoryContext = MemoryContextSwitchTo(CacheMemoryContext);
+
+	availableVersionCompatible = palloc0(sizeof(bool));
+	*availableVersionCompatible = MajorVersionsCompatible(availableVersion,
```
A pointer here is pretty awkward - why don't you instead just redo the check as long as availableVersionCompatible == false, and set it to true the first time it passes?

byucesoy replied:
Makes sense.

anarazel commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -1070,16 +1116,59 @@ CitusHasBeenLoaded(void)
+								errhint("Run ALTER EXTENSION citus UPDATE and try "
+										"again.")));
+			}
+		}
```
Still think a single caching variable (like `bool citusVersionsCompatible`) would be better.  Set to false by default and when extension dropped, set to true when version matches.  As long as it's false do the expensive check, otherwise the slow stuff.

byucesoy replied:
We show different error messages based on which version differs from loaded libraries' version. I am not sure whether we should combine them(?)

anarazel replied:
What I'm saying is that you shouldn't individually cache the results of the different relevant versions, but instead have a single cache: citusVersionsCompatible. If that's false, always perform both checks in an uncached manner.

byucesoy replied:
I think I got it. It is late here and my cognitive abilities are non-existent :smile:  It is best if I continue this tomorow.

byucesoy commented:
I also go forward and add checks for almost all of our UDFs for version mismatch. These UDFs should have been call CitusHasBeenLoaded. Was there a specific reson to not call it? I added this change as separate commit so we can drop if there is a reason that I am not aware of.

To not miss any UDF, I scanned citus--7.0-1.sql file, and here are some UDFs I did not add the checks explicitly;

No Citus metadata access (As far as I realized):
* citus_extradata_container
* column_name_to_column
* column_to_column_name
* lock_shard_metadata
* lock_shard_resources
* shard_name
* worker_find_block_local_path
* worker_foreign_file_path

Not C functions (What should we do with them?):
* citus_drop_trigger
* citus_truncate_trigger
* create_insert_proxy_for_table
* distributed_tables_colocated
* run_command_on_colocated_placements
* run_command_on_placements
* run_command_on_shards
* run_command_on_workers

Enterprise features (We should open separate PR in enterprise for these):
* isolate_tenant_to_new_shard
* master_move_shard_placement

I added checks to these, but I am not sure:
* master_dist_local_group_cache_invalidate
* master_dist_node_cache_invalidate
* master_dist_partition_cache_invalidate
* master_dist_placement_cache_invalidate
* master_dist_shard_cache_invalidate

anarazel commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1474,8 +1422,8 @@ ErrorIfUnstableCreateOrAlterExtensionStmt(Node *parsetree)
-		availableExtensionVersion = NULL;
-		ErrorIfAvailableVersionMismatch();
+		citusVersionKnownCompatible = false;
+		CheckCitusVersion(ERROR);
```
I guess this is intended to protect against somebody installing a new version while connected, and then trying to perform an update?

jasonmp85 replied:
Yeah, that was the intent of the _in case a new default is available_ comment.

byucesoy replied:
Yes, basically we do not want users to run ALTER EXTENSION UPDATE before restarting the database first.

anarazel commented on src/backend/distributed/worker/worker_merge_protocol.c:
```diff
@@ -81,6 +81,7 @@ worker_merge_files_into_table(PG_FUNCTION_ARGS)
 	/* we should have the same number of column names and types */
 	int32 columnNameCount = ArrayObjectCount(columnNameObject);
 	int32 columnTypeCount = ArrayObjectCount(columnTypeObject);
+
```
Not a problem, but why was this line changed?

byucesoy replied:
Oops, it remained from a change I revert back.  I will fix this with the test change.

anarazel commented on src/backend/distributed/worker/worker_partition_protocol.c:
```diff
@@ -112,6 +112,7 @@ worker_range_partition_table(PG_FUNCTION_ARGS)
 
 	/* first check that array element's and partition column's types match */
 	Oid splitPointType = ARR_ELEMTYPE(splitPointObject);
+
```
Not a problem, but why was this line changed?

anarazel commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -78,6 +78,71 @@ ALTER EXTENSION citus UPDATE TO '6.1-15';
+        ) shard_sizes;
+
+        RETURN table_size;
+END;
```
@byucesoy What does this test, and whats the benefit of actually having a meaningful function body?

byucesoy replied:
It causes next ALTER EXTENSION citus UPDATE TO '6.2-2'; fail. Because we also create a function with same name. It tests we can continue to operate when altering extension failed. I added it because of [this issue](https://github.com/citusdata/citus/issues/1422).

jasonmp85 replied:
Yeah, it's a dummy test function… we should excise the body.

anarazel commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -78,6 +78,71 @@ ALTER EXTENSION citus UPDATE TO '6.1-15';
 ALTER EXTENSION citus UPDATE TO '6.1-17';
 ALTER EXTENSION citus UPDATE TO '6.2-1';
+
+-- Test non-distributed queries work even in version mismatch
```
Not sure I like this being part of the series of all update steps here..., but I don't feel strongly about it.

jasonmp85 replied:
I'm fine with it, though after this patch goes in maybe it can live in my version tester matrix instead of here.

jasonmp85 replied:
(in fact that's where I think it should live after this is released, otherwise this'll be a bit weird as more version upgrade scripts are added and whatnot)

byucesoy replied:
I put the these tests after all upgrade steps.

jasonmp85 commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -78,6 +78,71 @@ ALTER EXTENSION citus UPDATE TO '6.1-15';
+-- Test SELECT
+SELECT * FROM version_mismatch_table ORDER BY column1;
+
+-- Test SELECT from pg_catalog
```
Can you issue a `\l` here instead? That's what causes problems with `pg_ctlcluster`, so maybe having it explicitly in the test could be good.

byucesoy replied:
\l also prints collation and ctype information which is likely to be change among different systems. (For example my laptop uses tr_TR.UTF-8 but our servers uses en_US.UTF-8). Since you said you have no blocking feedback, I am skipping this one.

jasonmp85 commented:
This change looks good to me; tested it some over the weekend and didn't
have any trouble. I have no blocking feedback for :shipit: so feel free to
test more over in Istanbul.

--
*Jason Petersen*
Software Engineer | Citus Data
303.736.9255
jason@citusdata.com

On May 19, 2017, at 5:22 PM, Burak Yücesoy <notifications@github.com> wrote:

*@byucesoy* commented on this pull request.
------------------------------

In src/backend/distributed/worker/worker_merge_protocol.c
<https://github.com/citusdata/citus/pull/1425#discussion_r117592108>:

> @@ -81,6 +81,7 @@ worker_merge_files_into_table(PG_FUNCTION_ARGS)
 	/* we should have the same number of column names and types */
 	int32 columnNameCount = ArrayObjectCount(columnNameObject);
 	int32 columnTypeCount = ArrayObjectCount(columnTypeObject);
+

Oops, it remained from a change I revert back. I will fix this with the
test change.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/citusdata/citus/pull/1425#discussion_r117592108>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AADF8UtwEqb0XpxTR-u0Z_4-Hhhqkkrvks5r7iQkgaJpZM4NfzAZ>
.

jasonmp85 commented:
> Enterprise features (We should open separate PR in enterprise for these):
> * isolate_tenant_to_new_shard
> * master_move_shard_placement

Uh, did we?
